libsawtooth/src/journal/genesis.rs

```diff
@@ -0,0 +1,638 @@
+        let observers = self.observers.ok_or_else(|| {
+            GenesisControllerBuildError::MissingField("'observers' field is required".to_string())
+        })?;
+
+        let initial_state_root = self.initial_state_root.ok_or_else(|| {
+            GenesisControllerBuildError::MissingField(
+                "'initial_state_root' field is required".to_string(),
+            )
+        })?;
+
+        let merkle_state = self.merkle_state.ok_or_else(|| {
+            GenesisControllerBuildError::MissingField(
+                "'merkle_state' field is required".to_string(),
+            )
+        })?;
+
+        let identity_signer = self.identity_signer.ok_or_else(|| {
+            GenesisControllerBuildError::MissingField(
+                "'identity_signer' field is required".to_string(),
+            )
+        })?;
+
+        let mut genesis_path_buf = PathBuf::from(data_dir);
+        genesis_path_buf.push(GENESIS_FILE);
+
+        Ok(GenesisController {
+            transaction_executor,
+            scheduler_factory,
+            block_manager,
+            chain_reader,
+            state_view_factory,
+            identity_signer,
+            chain_id_manager,
+            observers,
+            genesis_file_path: genesis_path_buf,
+            initial_state_root,
+            merkle_state,
+        })
+    }
+}
+
+/// The `GenesisController` is in charge of checking if this is the genesis node and creating the
+/// the genesis block from the genesis batch file
+pub struct GenesisController {
+    transaction_executor: ExecutionTaskSubmitter,
+    scheduler_factory: Box<dyn SchedulerFactory>,
+    block_manager: BlockManager,
+    chain_reader: Box<dyn ChainReader>,
+    state_view_factory: StateViewFactory,
+    identity_signer: Box<dyn Signer>,
+    chain_id_manager: ChainIdManager,
+    observers: Vec<Box<dyn ChainObserver>>,
+    initial_state_root: String,
+    merkle_state: CborMerkleState,
+    genesis_file_path: PathBuf,
+}
+
+impl GenesisController {
+    /// Determines if the system should be put in genesis mode
+    ///
+    /// Returns:
+    ///      bool: return whether or not a genesis block is required to be
+    ///
+    /// Returns error if there is invalid combination of the following: genesis.batch, existing
+    /// chain head, and block chain id.
+    pub fn requires_genesis(&self) -> Result<bool, GenesisError> {
+        // check if the genesis batch file exists
+        let genesis_path = self.genesis_file_path.as_path();
+        if genesis_path.is_file() {
+            debug!("Genesis batch file: {}", self.genesis_file_path.display());
+        } else {
+            debug!("Genesis batch file: not found",);
+        }
+
+        // check if there is a chain head
+        let chain_head = self.chain_reader.chain_head().map_err(|err| {
+            GenesisError::InvalidGenesisState(format!("Unable to read chain head: {}", err))
+        })?;
+
+        if chain_head.is_some() && genesis_path.is_file() {
```

Comment:
Can we avoid the expect with something like "if let Some(chain_head_inner) = chain_head && genesis_path.is_file() {"?

Reply:
Currently, combining `if let` with another condition is experimental (I tried this originally) https://github.com/rust-lang/rust/issues/53667

Reply:
How about:

if let Some(chain_head_inner) = chain_head {
    if genesis_path.is_file() {

Reply:
looks like add `ref` to if `let Some(ref chain_head_inner) = chain_head {` works